File: Data-Analysis/Regression/regression.py
import numpy as np
import statsmodels.formula.api as sfm

def backward_elimination(x, y, sl):
    '''
    :param x: array, independent variables
    :param y: target variable
    :param sl: significant level
    :return: array of optimised feature variables
    '''
    num_vars = len(x[0])
    for i in range(0, num_vars):
        result = sfm.OLS(y, x).fit()
        max_var = max(result.pvalues).astype(float)
        if max_var > sl:
            x = np.delete(x, np.argmax(result.pvalues), 1)
    result.summary()
    return x

File: Data-Analysis/Regression/regression.R
# backward elimination
backwardElimination <- function(x, sl) {
    numVars = length(x)
    for (i in c(1:numVars)){
      regressor = lm(formula = Profit ~ ., data = x)
      maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
      if (maxVar > sl){
        j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
        x = x[, -j]
      }
      numVars = numVars - 1
    }
    return(summary(regressor))
  }


sl = 0.05
df = df[, c(1, 2, 3, 4, 5)]
backward_elimination(df, sl)
